autogluon/utils/tabular/ml/models/tab_transformer/TabTransformer_model.py

```diff
@@ -0,0 +1,320 @@
+        self.fe=fe
+        if self.fe is not None:
+            if X_unlabeled is None:
+                unlab_data=None
+            elif X_unlabeled is not None:
+                unlab_data = utils.TabTransformerDataset(X_unlabeled, col_info=self.types_of_features, **self.kwargs)
+        if self.fe is None:
+            if X_unlabeled is None:
+                data.fit_feat_encoders()
+                self.fe = data.feature_encoders
+                unlab_data=None
+            elif X_unlabeled is not None:
+                unlab_data = utils.TabTransformerDataset(X_unlabeled, col_info=self.types_of_features, **self.kwargs)
+                unlab_data.fit_feat_encoders()
+                self.fe = unlab_data.feature_encoders
+
+
+        data.encode(self.fe)
+
+        if X_val is not None:
+            val_data = utils.TabTransformerDataset(X_val, col_info=self.types_of_features, **self.kwargs)
+            val_data.encode(self.fe)
+        else:
+            val_data=None
+
+        if unlab_data is not None:
+            unlab_data.encode(self.fe)
+
+        return data, val_data, unlab_data
+
+
+
+    def _fit(self, X_train, y_train, X_val=None, y_val=None, X_unlabeled=None, **kwargs):
+        self.set_default_params(y_train)
+
+        train, val, unlab = self.preprocess(X_train, X_val, X_unlabeled)
+   
+        if self.problem_type=='regression':
+            train.targets = torch.FloatTensor(list(y_train))
+            val.targets   = torch.FloatTensor(list(y_val))
+        else:
+            train.targets = torch.LongTensor(list(y_train))
+            val.targets   = torch.LongTensor(list(y_val))
+
+        trainloader, valloader, unlabloader = train.build_loader(), val.build_loader(), unlab.build_loader() if unlab is not None else None
+   
+        self.cat_feat_origin_cards=trainloader.cat_feat_origin_cards
+       
+        self.get_model()
+
+        if X_unlabeled is not None:
+            self.model.fit(unlabloader, valloader, state='pretrain') 
+            self.model.fit(trainloader, valloader, state='finetune')
+        else:
+            self.model.fit(trainloader, valloader)
+
+        """
+        #TODO: If no X_train is passed, but X_unlabeled is, assume that the user wants to 
+        pretrain and save a base model. (see todo list at bottom of script for more info)
+        """
+         
+
+    def predict_proba(self, X, preprocess=False):
+        """
+        X (torch.tensor or pd.dataframe): data for model to give prediction probabilities
+        returns: np.array of k-probabilities for each of the k classes. If k=2 we drop the second probability.
+        """
+        if preprocess or isinstance(X, pd.DataFrame):
+            X, _, _ =self.preprocess(X, fe=self.fe)
+        else:
+            X=X
+     
+        self.model.eval()
+        softmax=nn.Softmax(dim=1)
+        with torch.no_grad():
+            out, _ = self.model(X.cat_data)
```

Comment:
Doesn't this need to be done in mini-batches?  What if `X` contains tons of rows?

Reply:
This should be addressed in the latest commit. Pre-allocated tensors and then fill them batch by batch from a loader.